File: OpenFlix-tvOS/OpenFlix-tvOS/Models/DTOs/MediaDTOs.swift
import Foundation

// MARK: - Library Sections

struct LibrarySectionsResponse: Codable {
    let MediaContainer: LibrarySectionsContainer?

    // Alternative key mapping for different server versions
    enum CodingKeys: String, CodingKey {
        case MediaContainer
    }
}

struct LibrarySectionsContainer: Codable {
    let size: Int?
    let Directory: [LibrarySectionDTO]?
    let directories: [LibrarySectionDTO]?

    // Support both key formats
    var allDirectories: [LibrarySectionDTO] {
        Directory ?? directories ?? []
    }
}

struct LibrarySectionDTO: Codable {
    let key: String
    let type: String
    let title: String
    let agent: String?
    let scanner: String?
    let language: String?
    let uuid: String?
    let updatedAt: Int?
    let scannedAt: Int?
    let createdAt: Int?
    let content: Bool?
    let directory: Bool?
    let contentChangedAt: Int?
    let hidden: Int?
    let count: Int?

    var id: Int {
        Int(key) ?? 0
    }
}

// MARK: - Media Container (Generic)

struct MediaContainerResponse: Codable {
    let MediaContainer: MediaContainer?
}

struct MediaContainer: Codable {
    let size: Int?
    let totalSize: Int?
    let offset: Int?
    let allowSync: Bool?
    let identifier: String?
    let librarySectionID: Int?
    let librarySectionTitle: String?
    let librarySectionUUID: String?
    let Metadata: [MediaItemDTO]?
    let Hub: [HubDTO]?
    let Directory: [LibrarySectionDTO]?
}

// MARK: - Media Item

struct MediaItemDTO: Codable {
    let ratingKeyValue: StringOrInt?
    let key: String?
    let guid: String?
    let type: String?
    let title: String?
    let originalTitle: String?
    let tagline: String?
    let summary: String?
    let thumb: String?
    let art: String?
    let banner: String?
    let year: Int?
    let duration: Int?
    let viewOffset: Int?
    let viewCount: Int?
    let contentRating: String?
    let audienceRating: Double?
    let rating: Double?
    let studio: String?
    let addedAt: Int?
    let updatedAt: Int?
    let originallyAvailableAt: String?
    let leafCount: Int?          // For shows: total episodes
    let viewedLeafCount: Int?    // For shows: watched episodes
    let childCount: Int?         // For shows: season count
    let index: Int?              // Season/episode number
    let parentIndex: Int?        // Season number for episodes
    let parentRatingKey: StringOrInt? // Parent media key
    let parentTitle: String?     // Parent title (show name for episodes)
    let parentThumb: String?
    let grandparentRatingKey: StringOrInt?
    let grandparentTitle: String?
    let grandparentThumb: String?
    let grandparentArt: String?
    let librarySectionID: Int?
    let librarySectionTitle: String?
    let Genre: [GenreDTO]?
    let Role: [RoleDTO]?
    let Director: [DirectorDTO]?
    let Writer: [WriterDTO]?
    let Country: [CountryDTO]?
    let Media: [MediaVersionDTO]?

    enum CodingKeys: String, CodingKey {
        case ratingKeyValue = "ratingKey"
        case key, guid, type, title, originalTitle, tagline, summary
        case thumb, art, banner, year, duration, viewOffset, viewCount
        case contentRating, audienceRating, rating, studio
        case addedAt, updatedAt, originallyAvailableAt
        case leafCount, viewedLeafCount, childCount
        case index, parentIndex, parentRatingKey, parentTitle, parentThumb
        case grandparentRatingKey, grandparentTitle, grandparentThumb, grandparentArt
        case librarySectionID, librarySectionTitle
        case Genre, Role, Director, Writer, Country, Media
    }

    // Computed properties with safe defaults
    var ratingKey: String? {
        ratingKeyValue?.stringValue
    }

    var ratingKeyInt: Int {
        ratingKeyValue?.intValue ?? 0
    }

    var safeTitle: String {
        title ?? "Unknown"
    }

    var safeType: String {
        type ?? "unknown"
    }

    var safeKey: String {
        key ?? ""
    }
}

// Helper type to handle fields that can be String or Int
enum StringOrInt: Codable {
    case string(String)
    case int(Int)

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let intValue = try? container.decode(Int.self) {
            self = .int(intValue)
        } else if let stringValue = try? container.decode(String.self) {
            self = .string(stringValue)
        } else {
            throw DecodingError.typeMismatch(StringOrInt.self, DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Expected String or Int"))
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let value):
            try container.encode(value)
        case .int(let value):
            try container.encode(value)
        }
    }

    var stringValue: String {
        switch self {
        case .string(let value): return value
        case .int(let value): return String(value)
        }
    }

    var intValue: Int {
        switch self {
        case .string(let value): return Int(value) ?? 0
        case .int(let value): return value
        }
    }
}

struct GenreDTO: Codable {
    let id: Int?
    let tag: String
}

struct RoleDTO: Codable {
    let id: Int?
    let tag: String
    let role: String?
    let thumb: String?
}

struct DirectorDTO: Codable {
    let id: Int?
    let tag: String
}

struct WriterDTO: Codable {
    let id: Int?
    let tag: String
}

struct CountryDTO: Codable {
    let id: Int?
    let tag: String
}

// MARK: - Media Versions & Streams

struct MediaVersionDTO: Codable {
    let id: Int?
    let duration: Int?
    let bitrate: Int?
    let width: Int?
    let height: Int?
    let aspectRatio: Double?
    let audioChannels: Int?
    let audioCodec: String?
    let videoCodec: String?
    let videoResolution: String?
    let container: String?
    let videoFrameRate: String?
    let Part: [MediaPartDTO]?

    var safeId: Int { id ?? 0 }
}

struct MediaPartDTO: Codable {
    let id: Int?
    let key: String?
    let duration: Int?
    let file: String?
    let size: Int?
    let container: String?
    let Stream: [StreamDTO]?

    var safeId: Int { id ?? 0 }
    var safeKey: String { key ?? "" }
}

struct StreamDTO: Codable {
    let id: Int?
    let streamType: Int?       // 1=video, 2=audio, 3=subtitle
    let codec: String?
    let index: Int?
    let language: String?
    let languageCode: String?
    let displayTitle: String?
    let selected: Bool?
    let forced: Bool?
    let `default`: Bool?
    let title: String?
    // Video-specific
    let width: Int?
    let height: Int?
    let bitrate: Int?
    let frameRate: Double?
    // Audio-specific
    let channels: Int?
    let samplingRate: Int?

    var safeId: Int { id ?? 0 }
    var safeStreamType: Int { streamType ?? 1 }
}

// MARK: - Hubs

struct HubsResponse: Codable {
    let MediaContainer: HubsContainer?
}

struct HubsContainer: Codable {
    let size: Int?
    let librarySectionID: Int?
    let Hub: [HubDTO]?
}

struct HubDTO: Codable {
    let key: String?
    let hubKey: String?
    let type: String?
    let hubIdentifier: String?
    let title: String?
    let context: String?
    let size: Int?
    let more: Bool?
    let style: String?
    let promoted: Bool?
    let Metadata: [MediaItemDTO]?

    var safeType: String { type ?? "unknown" }
    var safeTitle: String { title ?? "" }
}

// MARK: - Streaming Services

struct StreamingServicesResponse: Codable {
    let MediaContainer: StreamingServicesContainer?
}

struct StreamingServicesContainer: Codable {
    let Directory: [StreamingServiceDTO]?
}

struct StreamingServiceDTO: Codable {
    let key: String?
    let title: String?
    let thumb: String?

    var safeKey: String { key ?? "" }
    var safeTitle: String { title ?? "" }
}

// MARK: - Search

struct SearchResponse: Codable {
    let MediaContainer: SearchContainer?
}

struct SearchContainer: Codable {
    let Hub: [SearchHubDTO]?
}

struct SearchHubDTO: Codable {
    let type: String?
    let title: String?
    let size: Int?
    let Metadata: [MediaItemDTO]?

    var safeType: String { type ?? "unknown" }
    var safeTitle: String { title ?? "" }
}

// MARK: - Playback

struct PlaybackURLResponse: Codable {
    let url: String
    let `protocol`: String?
    let directPlay: Bool?
    let transcoding: Bool?

    enum CodingKeys: String, CodingKey {
        case url
        case `protocol`
        case directPlay = "direct_play"
        case transcoding
    }
}

// MARK: - Filters & Sorts

struct FiltersResponse: Codable {
    let MediaContainer: FiltersContainer
}

struct FiltersContainer: Codable {
    let filterTypes: [FilterTypeDTO]?

    enum CodingKeys: String, CodingKey {
        case filterTypes = "Type"
    }
}

struct FilterTypeDTO: Codable {
    let key: String
    let type: String
    let title: String
    let Filter: [FilterDTO]?
}

struct FilterDTO: Codable {
    let filter: String
    let filterType: String
    let title: String
    let type: String
}

struct SortsResponse: Codable {
    let MediaContainer: SortsContainer
}

struct SortsContainer: Codable {
    let FieldType: [SortFieldDTO]?
}

struct SortFieldDTO: Codable {
    let key: String
    let title: String
    let type: String?
}

File: android-native/app/src/main/java/com/openflix/data/remote/dto/MediaDtos.kt
package com.openflix.data.remote.dto

import com.google.gson.annotations.SerializedName

/**
 * Media-related DTOs
 */

data class MediaItemDto(
    @SerializedName("ratingKey") val ratingKey: Int?,  // Plex uses integer ratingKey as primary ID
    @SerializedName("key") val key: String?,
    @SerializedName("guid") val guid: String?,
    @SerializedName("type") val type: String = "unknown",  // movie, show, season, episode, clip
    @SerializedName("title") val title: String = "",
    @SerializedName("originalTitle") val originalTitle: String?,
    @SerializedName("tagline") val tagline: String?,
    @SerializedName("summary") val summary: String?,
    @SerializedName("thumb") val thumb: String?,
    @SerializedName("art") val art: String?,
    @SerializedName("banner") val banner: String?,
    @SerializedName("year") val year: Int?,
    @SerializedName("duration") val duration: Long?,  // milliseconds
    @SerializedName("viewOffset") val viewOffset: Long?,  // watch progress in milliseconds
    @SerializedName("viewCount") val viewCount: Int?,
    @SerializedName("addedAt") val addedAt: Long?,
    @SerializedName("updatedAt") val updatedAt: Long?,
    @SerializedName("originallyAvailableAt") val originallyAvailableAt: String?,
    @SerializedName("contentRating") val contentRating: String?,
    @SerializedName("rating") val rating: Double?,
    @SerializedName("audienceRating") val audienceRating: Double?,
    @SerializedName("studio") val studio: String?,
    @SerializedName("Genre") val genres: List<GenreDto>?,
    @SerializedName("Director") val directors: List<PersonDto>?,
    @SerializedName("Writer") val writers: List<PersonDto>?,
    @SerializedName("Role") val roles: List<RoleDto>?,
    @SerializedName("Media") val media: List<MediaVersionDto>?,
    @SerializedName("librarySectionID") val librarySectionId: Int?,
    @SerializedName("librarySectionTitle") val librarySectionTitle: String?,

    // For TV shows
    @SerializedName("parentRatingKey") val parentRatingKey: Int?,
    @SerializedName("parentTitle") val parentTitle: String?,
    @SerializedName("parentThumb") val parentThumb: String?,
    @SerializedName("grandparentRatingKey") val grandparentRatingKey: Int?,
    @SerializedName("grandparentTitle") val grandparentTitle: String?,
    @SerializedName("grandparentThumb") val grandparentThumb: String?,
    @SerializedName("grandparentArt") val grandparentArt: String?,
    @SerializedName("index") val index: Int?,  // episode/season number
    @SerializedName("parentIndex") val parentIndex: Int?,  // season number for episodes
    @SerializedName("leafCount") val leafCount: Int?,  // episode count
    @SerializedName("viewedLeafCount") val viewedLeafCount: Int?,
    @SerializedName("childCount") val childCount: Int?  // season count for shows
) {
    // Helper to get ID as string (uses ratingKey)
    val id: String get() = ratingKey?.toString() ?: guid ?: ""
}

data class MediaVersionDto(
    @SerializedName("id") val id: String?,
    @SerializedName("duration") val duration: Long?,
    @SerializedName("bitrate") val bitrate: Int?,
    @SerializedName("width") val width: Int?,
    @SerializedName("height") val height: Int?,
    @SerializedName("aspect_ratio") val aspectRatio: Double?,
    @SerializedName("audio_channels") val audioChannels: Int?,
    @SerializedName("audio_codec") val audioCodec: String?,
    @SerializedName("video_codec") val videoCodec: String?,
    @SerializedName("video_resolution") val videoResolution: String?,
    @SerializedName("container") val container: String?,
    @SerializedName("video_frame_rate") val videoFrameRate: String?,
    @SerializedName("video_profile") val videoProfile: String?,
    @SerializedName("Part") val part: List<MediaPartDto>?
)

data class MediaPartDto(
    @SerializedName("id") val id: String?,
    @SerializedName("key") val key: String?,
    @SerializedName("duration") val duration: Long?,
    @SerializedName("file") val file: String?,
    @SerializedName("size") val size: Long?,
    @SerializedName("container") val container: String?,
    @SerializedName("streams") val streams: List<MediaStreamDto>?
)

data class MediaStreamDto(
    @SerializedName("id") val id: String?,
    @SerializedName("stream_type") val streamType: Int?,  // 1=video, 2=audio, 3=subtitle
    @SerializedName("index") val index: Int?,
    @SerializedName("codec") val codec: String?,
    @SerializedName("language") val language: String?,
    @SerializedName("language_code") val languageCode: String?,
    @SerializedName("title") val title: String?,
    @SerializedName("display_title") val displayTitle: String?,
    @SerializedName("selected") val selected: Boolean?,
    @SerializedName("default") val default: Boolean?,
    @SerializedName("forced") val forced: Boolean?,

    // Video specific
    @SerializedName("width") val width: Int?,
    @SerializedName("height") val height: Int?,
    @SerializedName("bit_depth") val bitDepth: Int?,
    @SerializedName("frame_rate") val frameRate: Double?,

    // Audio specific
    @SerializedName("channels") val channels: Int?,
    @SerializedName("sampling_rate") val samplingRate: Int?
)

data class GenreDto(
    @SerializedName("id") val id: String?,
    @SerializedName("tag") val tag: String
)

data class PersonDto(
    @SerializedName("id") val id: String?,
    @SerializedName("tag") val tag: String,
    @SerializedName("thumb") val thumb: String?
)

data class RoleDto(
    @SerializedName("id") val id: String?,
    @SerializedName("tag") val tag: String,  // actor name
    @SerializedName("role") val role: String?,  // character name
    @SerializedName("thumb") val thumb: String?
)

data class MediaListResponse(
    @SerializedName("items") val items: List<MediaItemDto>,
    @SerializedName("total_size") val totalSize: Int?,
    @SerializedName("offset") val offset: Int?,
    @SerializedName("size") val size: Int?
)

data class SeasonDto(
    @SerializedName("id") val id: String,
    @SerializedName("rating_key") val ratingKey: String?,
    @SerializedName("title") val title: String,
    @SerializedName("index") val index: Int,
    @SerializedName("thumb") val thumb: String?,
    @SerializedName("leaf_count") val leafCount: Int?,
    @SerializedName("viewed_leaf_count") val viewedLeafCount: Int?,
    @SerializedName("parent_rating_key") val parentRatingKey: String?
)

data class EpisodeDto(
    @SerializedName("id") val id: String,
    @SerializedName("rating_key") val ratingKey: String?,
    @SerializedName("title") val title: String,
    @SerializedName("index") val index: Int,
    @SerializedName("parent_index") val parentIndex: Int?,
    @SerializedName("summary") val summary: String?,
    @SerializedName("thumb") val thumb: String?,
    @SerializedName("duration") val duration: Long?,
    @SerializedName("view_offset") val viewOffset: Long?,
    @SerializedName("originally_available_at") val originallyAvailableAt: String?,
    @SerializedName("parent_rating_key") val parentRatingKey: String?,
    @SerializedName("grandparent_rating_key") val grandparentRatingKey: String?
)

data class PlaybackResponse(
    @SerializedName("url") val url: String,
    @SerializedName("protocol") val protocol: String?,
    @SerializedName("direct_play") val directPlay: Boolean?,
    @SerializedName("transcoding") val transcoding: Boolean?
)

data class ProgressUpdateRequest(
    @SerializedName("key") val key: String,
    @SerializedName("time") val time: Long,  // milliseconds
    @SerializedName("duration") val duration: Long?,
    @SerializedName("state") val state: String?  // playing, paused, stopped
)

data class SearchResponse(
    @SerializedName("MediaContainer") val mediaContainer: SearchMediaContainer?
)

data class SearchMediaContainer(
    @SerializedName("Hub") val hubs: List<SearchHub>?
)

data class SearchHub(
    @SerializedName("type") val type: String?,
    @SerializedName("title") val title: String?,
    @SerializedName("Metadata") val metadata: List<MediaItemDto>?
)

// === Plex-compatible DTOs ===

/**
 * Response wrapper for Plex-compatible MediaContainer
 */
data class MediaContainerResponse(
    @SerializedName("MediaContainer") val mediaContainer: MediaContainer?
)

data class MediaContainer(
    @SerializedName("size") val size: Int?,
    @SerializedName("totalSize") val totalSize: Int?,
    @SerializedName("offset") val offset: Int?,
    @SerializedName("Metadata") val metadata: List<MediaItemDto>?,
    @SerializedName("Directory") val directory: List<DirectoryDto>?
)

/**
 * Library sections response
 */
data class LibrarySectionsResponse(
    @SerializedName("MediaContainer") val mediaContainer: LibrarySectionsContainer?
)

data class LibrarySectionsContainer(
    @SerializedName("size") val size: Int?,
    @SerializedName("Directory") val directories: List<LibrarySectionDto>?
)

data class LibrarySectionDto(
    @SerializedName("key") val key: String,
    @SerializedName("type") val type: String,  // movie, show
    @SerializedName("title") val title: String,
    @SerializedName("agent") val agent: String?,
    @SerializedName("scanner") val scanner: String?,
    @SerializedName("language") val language: String?,
    @SerializedName("uuid") val uuid: String?,
    @SerializedName("count") val count: Int?,
    @SerializedName("scannedAt") val scannedAt: Long?,
    @SerializedName("createdAt") val createdAt: Long?,
    @SerializedName("updatedAt") val updatedAt: Long?
)

data class DirectoryDto(
    @SerializedName("key") val key: String?,
    @SerializedName("title") val title: String?,
    @SerializedName("type") val type: String?
)

// === Hubs (Recommendations) DTOs ===

data class HubsResponse(
    @SerializedName("MediaContainer") val mediaContainer: HubsContainer?
)

data class HubsContainer(
    @SerializedName("size") val size: Int?,
    @SerializedName("librarySectionID") val librarySectionId: Int?,
    @SerializedName("Hub") val hubs: List<StreamingHubDto>?
)

data class StreamingHubDto(
    @SerializedName("key") val key: String?,
    @SerializedName("hubIdentifier") val hubIdentifier: String?,
    @SerializedName("type") val type: String?,
    @SerializedName("title") val title: String = "",
    @SerializedName("context") val context: String?,  // e.g., "hub.streamingService", "hub.genre"
    @SerializedName("size") val size: Int?,
    @SerializedName("more") val more: Boolean?,
    @SerializedName("style") val style: String?,
    @SerializedName("promoted") val promoted: Boolean?,
    @SerializedName("Metadata") val metadata: List<MediaItemDto>?
)
